src/java/org/apache/cassandra/dht/tokenallocator/IsolatedTokenAllocator.java

```diff
@@ -0,0 +1,156 @@
+        private final TokenAllocation allocation;
+        private final Map<String, SummaryStatistics> lastCheckPoint = Maps.newHashMap();
+
+        private QuietAllocator(AbstractReplicationStrategy source)
+        {
+            this.quietSnitch = new QuietSnitch();
+            this.quietTokenMetadata = source.getTokenMetadata().cloneWithNewSnitch(quietSnitch);
+            // TODO all or full replicas?
+            var rf = source.getReplicationFactor().allReplicas;
+            var numTokens = DatabaseDescriptor.getNumTokens();
+            this.allocation = TokenAllocation.create(quietSnitch, quietTokenMetadata, rf, numTokens);
+        }
+
+        private Collection<Token> allocateTokensForNode(int nodeId, String rackId)
+        {
+            // Update snitch and token metadata info to inform token allocation
+            InetAddressAndPort fakeNodeAddressAndPort = getLoopbackAddressWithPort(nodeId);
+            quietSnitch.nodeByRack.put(fakeNodeAddressAndPort, rackId);
+            quietTokenMetadata.updateTopology(fakeNodeAddressAndPort);
+
+            // Allocate tokens
+            Collection<Token> tokens = allocation.allocate(fakeNodeAddressAndPort);
+
+            // Validate ownership stats
+            validateAllocation(nodeId, rackId);
+
+            return tokens;
+        }
+
+        private void validateAllocation(int nodeId, String rackId)
+        {
+            SummaryStatistics newOwnership = allocation.getAllocationRingOwnership(SimpleSnitch.DATA_CENTER_NAME, rackId);
+            SummaryStatistics oldOwnership = lastCheckPoint.put(rackId, newOwnership);
+            if (oldOwnership != null)
+                logger.debug(String.format("Replicated node load in rack=%d before allocating node %d: %s.", rackId, nodeId,
+                                           TokenAllocation.statToString(oldOwnership)));
+            logger.debug(String.format("Replicated node load in rack=%d after allocating node %d: %s.", rackId, nodeId,
+                                       TokenAllocation.statToString(newOwnership)));
+            if (oldOwnership != null && oldOwnership.getStandardDeviation() != 0.0)
+            {
+                double stdDevGrowth = newOwnership.getStandardDeviation() - oldOwnership.getStandardDeviation();
+                if (stdDevGrowth > TokenAllocation.WARN_STDEV_GROWTH)
+                {
+                    logger.warn(String.format("Growth of %.2f%% in token ownership standard deviation after allocating node %d on rack %d above warning threshold of %d%%",
+                                              stdDevGrowth * 100, nodeId, rackId, (int)(TokenAllocation.WARN_STDEV_GROWTH * 100)));
+                }
+            }
+        }
+    }
+
+    private static class QuietSnitch extends SimpleSnitch
+    {
+        final Map<InetAddressAndPort, String> nodeByRack = new HashMap<>();
+
+        @Override
+        public String getRack(InetAddressAndPort endpoint)
```

Comment:
Will this return the correct rack for the preexisting nodes?

Reply:
Great catch, it won't. I need to update it to call both `nodeByRack` and the real topology for the cluster.